File: java/src/main/java/hubert/akka/Node.java
package hubert.akka;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.Props;

import java.util.ArrayList;

import hubert.akka.Master.NotFinished;
import hubert.akka.Master.UpdateEstimate;

public class Node extends AbstractActor {
  static public Props props() {
    return Props.create(Node.class, () -> new Node());
  }

  static public class Neighbours {
    public final ArrayList<ActorRef> neighs;

    public Neighbours(ArrayList<ActorRef> neighbours) {
      this.neighs = neighbours;
    }
  }

  static public class DistanceEstimate {
    public final int distance;

    public DistanceEstimate(Integer new_distance) {
      this.distance = new_distance;
    }
  }

  static public class AskDistance {
  };

  static public class Propagate {
  };

  private ArrayList<ActorRef> neigh;
  private int dist = Integer.MAX_VALUE;
  private int propagated = dist;
  private boolean has_propagated = true;
  private ActorRef supervisor;

  public Node() {
  }

  private void onDistanceEstimate(DistanceEstimate wtg) {
    if (wtg.distance >= this.dist)
      return;
    this.dist = wtg.distance;
    if (!this.has_propagated)
      return;
    // send the messages just for the first time, clear mailbox then
    this.supervisor.tell(new NotFinished(), getSelf());
    self().tell(new Propagate(), getSelf());
    this.has_propagated = false;
  }

  private void onPropagate() {
    for (ActorRef n : this.neigh) {
      n.tell(new DistanceEstimate(this.dist + 1), getSelf());
    }
    this.has_propagated = true;
    self().tell(new AskDistance(), getSelf());
  }

  private void onAskDistance() {
    if (has_propagated) // tell the supervisor only if no propagation is due
      supervisor.tell(new UpdateEstimate(this.dist), getSelf());
  }

  private void onNeighbours(Neighbours nb) {
    this.neigh = nb.neighs;
    this.supervisor = getSender();
  }

  @Override
  public Receive createReceive() {
    return receiveBuilder() //
        .match(DistanceEstimate.class, wtg -> onDistanceEstimate(wtg)) //
        .match(Propagate.class, mes -> onPropagate()) //
        .match(AskDistance.class, (mes) -> onAskDistance()) //
        .match(Neighbours.class, nb -> onNeighbours(nb)) //
        .matchEquals("show neighbours", mes -> {
          getSender().tell(neigh, getSelf());
        }).build();
  }
}

File: java/src/main/scala/hubert/akka/Node.scala
package hubert.akka

import java.util.ArrayList

import akka.actor.{Actor, ActorLogging, ActorRef, ActorSystem, Props}

//remove if not needed
import scala.collection.JavaConversions._

class Status(dist: Double) { //with SendingCorrections
  var distance = dist
  var announced = 0.0
  var propagated = false
  var newMessages = false
  var sentReady = true
  def getCorrection(): Double = {
    val ret = distance - announced
    announced = distance
    return ret
  }
}

object Node {

  def props: Props = Props(new Node())

  final case class Neighbours(neighbours: Array[ActorRef])

  final case class DistanceEstimate(distance: Double, src: ActorRef)

  final case class NewSource(sourceId: ActorRef)

  final case class UpdateParent(source: ActorRef)

  final case class Propagate(soruce: ActorRef)

}

class Node() extends Actor with ActorLogging {
  import hubert.akka.Master.{NotIdle, CorrectBy}
  import Node._

  private var neigh: Array[ActorRef] = _
  private var active: Map[ActorRef, Status] = _

  def onDistanceEstimate(distance: Double, src: ActorRef): Unit = {
    if (! active.contains(src)) { // check this against docs
      active += (src -> new Status(distance))
      self ! UpdateParent(src)
      self ! Propagate(src)
    } else {
      var status = active(src)
      status.newMessages = true
      if (status.distance > distance) {
        if(status.sentReady)
          context.parent ! NotIdle(src)  
        if (status.propagated)
          self ! Propagate(src)

        status.propagated = false
        status.distance = distance
      }
    }
  }

  def onPropagate(src: ActorRef): Unit = {
    if (! active.contains(src))
      log.info("Error, propagating inactive source")
    else {
      var status = active(src)
      for (n <- neigh) {
        n ! DistanceEstimate(status.distance + 1, src)
      }
      status.propagated = true
    }
  }

  def onUpdateParent(src: ActorRef) {
    if (! active.contains(src))
      log.info("Error, updating inactive source")
    else {
      var status = active(src)
      if (!status.newMessages) {
        context.parent ! CorrectBy(status.getCorrection(), src)
        // log.info("Updated {}", dist)
        status.sentReady = true
      } else {
        status.newMessages = false
        self ! UpdateParent(src)
      }
    }
  }

  def receive = {
    case Neighbours(nb)              => neigh = nb
    case DistanceEstimate(dist, src) => onDistanceEstimate(dist, src)
    case Propagate(src)              => onPropagate(src)
    case UpdateParent(src)           => onUpdateParent(src)

  }
}
